Ees 1407

luke-h1 opened the pull request:
This PR refactors publish_release_and_amend_2.robot to use more filters as stated in EES-1407. Instead of using Dates.csv as the initial file it now uses multiple_filters.csv which allows the test to utilize a lot more filters than dates.csv. Two new files have been added to the files/ directory (multiple_filters.csv & multiple_filters.meta.csv )

markhiveit commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -102,41 +102,68 @@ Select "Pupil Exclusions" publication
+    user clicks indicator checkbox   Unauthorised absence rate
+    user checks indicator checkbox is checked  Unauthorised absence rate
+
+Select gender filter
```
I'd add more filters to this - although not a crazy amount. To test a subject with multiple filters thoroughly, we want to select a few and then look at the data in the resulting table and check it is correct.

markhiveit commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -145,7 +172,7 @@ Create table
 Validate results table column headings
     [Tags]  HappyPath
-    user checks table column heading contains  1   1   2020 Week 14
+    user checks table column heading contains  1   1  Female
```
We should add more checks to the table itself, to check the data is correct - and cross reference what is being produced with the CSV file to make sure it is actually correct.

markhiveit commented on tests/robot-tests/tests/files/multiple_filters.meta.csv:
```diff
@@ -0,0 +1,10 @@
+sess_unauthorised_percent,Indicator,Unauthorised absence rate,Absence fields,£m,2,,
+enrolments_pa_10_exact,Indicator,Number of persistent absentees,Absence fields,,0,,
+enrolments_pa_10_exact_percent,Indicator,Percentage of persistent absentees,Absence fields,%,3,,
+school_type,Filter,School type,,,,,school_phase
```
This file only has two filters - I think we really want a subject with 5-6 filters.

luke-h1 commented:
Refactored test to check table after methodology change, added `SevenFilters` file in order for the test to utilize more filters and delete `create_amendment_check_permalink.robot `as this test contains the permalink checking functionality.

ntsim commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -237,6 +417,76 @@ Go to permalink page & check for error element to be present
+    user checks results table row heading contains  9  1   Black/African/Caribbean/Black British
+    user checks results table row heading contains  10  1  Mixed/Multiple ethnic group
+    user checks results table row heading contains  11  1  Not Known/Not Provided
+    user checks results table row heading contains  11  1  Not Known/Not Provided
```
This is repeated

ntsim commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -237,6 +417,76 @@ Go to permalink page & check for error element to be present
+    user checks results table row heading contains  35  1  White
+
+    # BLUE
+    user checks results table cell contains  1   1  n/a
```
Is the entire column n/a'd? Table tool should filter out columns where the entirety of it is n/a, so might be a bug

luke-h1 replied:
The majority of columns are n/a but there are some values at the start of stilton:  
![image](https://user-images.githubusercontent.com/55030296/99266526-c1c36180-281a-11eb-8c6b-bfd48d6be32f.png)

ntsim commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -102,50 +166,143 @@ Select "Pupil Exclusions" publication
+Validate table cells
+    [Tags]  HappyPath
+    # BLUE
+    user checks results table cell contains  1   1  n/a
```
Is the entire column n/a'd? Table tool should filter out columns where the entirety of it is n/a, so might be a bug

ntsim commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -8,25 +8,89 @@ Suite Setup       user signs in as bau1
 
 *** Variables ***
-${DETAILS_HEADING}  Academic Year 2020/21
+${DETAILS_HEADING}  Academic Year Q1 2020/21
```
`DETAILS_HEADING` is way too generic of a name (we use `details` components . It sounds like this should be `RELEASE_NAME` instead

ntsim commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -8,25 +8,89 @@ Suite Setup       user signs in as bau1
+    user waits until page contains button  New section
+    user clicks button  New section
+    user scrolls down  200
+    user waits until page contains button  Add text block
```
Don't think we need all these waits as `user clicks button` already does some waiting. Same applies to any other buttons you've clicked on in this test case.

Check out how we add content in the `release_content` suite to see how we've done it there.

ntsim commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -8,25 +8,89 @@ Suite Setup       user signs in as bau1
+    user clicks button  Add text block
+    user waits until page contains button  Edit block
+    user clicks button  Edit block
+    user scrolls down  200
```
Are we scrolling some more because of lazy loading? The content section should already be rendered by now if I'm not mistaken

luke-h1 replied:
Yep. Initially I was having lots of issues trying to find these buttons. Not sure if it was because they are lazy loaded or whether it was due to some quirk of selenium. Either way I've removed them now and the test seems to be ok now without them.

ntsim commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -161,14 +318,37 @@ Go to permalink
+    user clicks button  Add text block
+    user waits until page contains button  Edit block
+    user clicks button  Edit block
+    user scrolls down  200
```
Same as before, why are we doing so much scrolling?

ntsim commented on tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:
```diff
@@ -161,14 +318,37 @@ Go to permalink
+    user scrolls down  200
+    user waits until page contains button  Add text block
+    user clicks button  Add text block
+    user waits until page contains button  Edit block
```
Same as above, don't think we need all these waits

ntsim reviewed (changes requested):
This suite has a lot of commonality with `publish_data`. I wonder if we need `publish_data` anymore? My only problem would be that `publish_data` generates a table that has realistic looking data, whereas it seems like the seven filters file might not.

Not sure if you have any thoughts about this?

luke-h1 commented:
I think we should keep `publish_data.robot`. This test covers a lot of the bases that `publish_data.robot` does so if this test eventually uses a more realistic set of data then it can be merged into one test case to reduce repetition of similar tests.
